migrating doc draft (#169)
signalfx/splunk-otel-js: MIGRATING.md

# Migrate from the SignalFx Tracing Library for NodeJS

The Splunk Distribution of OpenTelemetry for NodeJS replaces the SignalFx Tracing
Library for NodeJS. If you’re using the SignalFx Tracing Library, migrate to

> the Splunk Distribution of OpenTelemetry for NodeJS to use OpenTelemetry’s

Reviewer: You can drop the 's

Because the SignalFx Tracing Library for NodeJS uses OpenTracing and the Splunk Distribution
of OpenTelemetry for NodeJS uses OpenTelemetry, the semantic
conventions for span tag names change when you migrate. For more information,
see [Migrate from OpenTracing to OpenTelemetry](https://docs.signalfx.com/en/latest/apm/apm-getting-started/apm-opentelemetry-collector.html#apm-opentelemetry-migration).

<a name="known-limitations"></a>
## Known limitations as compared to SignalFx Tracing Library

- Lowest supported version of NodeJS is `v8.5`, [see more information here](https://github.com/open-telemetry/opentelemetry-js#node-support)
- No auto-instrumentation for:
  - `AdonisJS`
  - `amqp10`
  - `mongodb-core` ([because it's deprecated](https://github.com/mongodb-js/mongodb-core))
  - `sails`
- Limited instrumentation for:
  - `nest` - only manual insturmentation helpers, provided by community
- other notes on instrumentation:
  - `express` instrumentation requires `http`/`https` instrumentation
  - `bluebird` - context propagation only (via `async_hooks`)
  - `q` - context propagation only (via `async_hooks`)
  - `when` - context propagation only (via `async_hooks`)

>   - `socket.io` - provided by community (<https://github.com/aspecto-io/opentelemetry-ext-js/tree/master/packages/instrumentation-socket.io>)

Reviewer: Not sure if we should talk of limitations. We want people to migrate, not to stick to the old stuff.

Can we change this section so that it states what the Splunk OTel dist for NodeJS is compatible with?

Author: Obviously we'll do our best to help the customers migrate, but let's make it clear, what the pitfalls are.

## Requirements

This Splunk Distribution of OpenTelemetry requires Node.js 8.5 or later.
If you're still using an earlier version of Node.js, continue using the SignalFx Tracing Library for Node.js.

> ## Equivalent configurations

Reviewer: ```suggestion
## Configurations you can migrate
```

### Changes to defaults

> - default flush interval is now 30s instead of 2s (how frequently captured telemetry data is sent to the backend)

Reviewer: ```suggestion
- Default flush interval, which defines how frequently captured telemetry data is sent to the backend, is now 30s instead of 2s 
```

### Instrumented libraries

With the exception of [explicitly listed limitations](#known-limitations) we aim to support all libraries supported by

> signalfx-nodejs-tracing. To find an equivalent auto-instrumentation open <https://opentelemetry.io/registry/> and for

Reviewer: Ditto. Consider removing mention of limitations.

each instrumentation
[from `signalfx-nodejs-tracing`'s README](https://github.com/signalfx/signalfx-nodejs-tracing/#requirements-and-supported-software)
search by the name of the library in the registry.

> For example, if you'd like to migrate instrumentation for `mysql`, go to

Reviewer: ```suggestion
For example, if you'd like to migrate instrumentation for MySQL, go to
```

Author: Well, actually, this is how the package is called, see <https://www.npmjs.com/package/mysql>. Maybe I should convert it into a link and then it'll be clearer?

[https://opentelemetry.io/registry/?s=**mysql**&component=&language=**js**#](https://opentelemetry.io/registry/?s=mysql&component=&language=js#).

> Once you have identified an instrumentation package, install it using `npm` or `yarn` and:

Reviewer: ```suggestion
Once you have identified an instrumentation package, install it using `npm` or `yarn`.
```

> - if the package is [on this list](./README.md#default-instrumentation-packages-) then it
>   will be enabled automatically
> - if it's not on the list, then please follow the steps for
>   [installing other instrumentation packages](./README.md#custom-instrumentation-packages)

Reviewer: ```suggestion
- If the package is [on this list](./README.md#default-instrumentation-packages-), it
  will be enabled automatically.
- if it's not on the list, follow the steps for
  [installing other instrumentation packages](./README.md#custom-instrumentation-packages).
```

| `logInjectionTags`       | -                       | no direct equivalent, but `tracerConfig.resource` can be used |
| `flushInterval`          | -                       | no direct equivalent, contact us if you had customized this value |
| `plugins`                | -                       | see [the README section about instrumentations](./README.md#custom-instrumentation-packages) |
| `recordedValueMaxLength` | `maxAttrLength`         | |
| `enableServerTiming`     | `serverTimingEnabled`   | |

### Invocation

```javascript
const tracer = require('signalfx-tracing').init({
  // your options here
})
```

becomes

```javascript
const { startTracing } = require('@splunk/otel');

startTracing({
  // your new options here
});
```

and requires installing `@splunk/otel` first, using either NPM or Yarn. Same as in `signalfx-nodejs-tracing`, this code
is to be run before your `import` or `require` statements.

Alternatively, you can append the flag `-r @splunk/otel/instrument` instead when launching `node` (it runs
`startTracing` under the hood). In that case you cannot use programmatic configuration and must rely on environment
variables only.

### Instrumentation logs

> There isn't a 1-to-1 mapping for `SIGNALFX_TRACING_DEBUG`. Closest equivalent is `OTEL_LOG_LEVEL`, however the logged
> information might be different. Please note that this section is about the logs **produced by instrumentation**, and not
> about the logs produced by the application.
> 

Reviewer: ```suggestion
There isn't a one-to-one mapping for `SIGNALFX_TRACING_DEBUG`. The closest equivalent is `OTEL_LOG_LEVEL`, however the logged
information might be different. 

> Note that this section is about the logs produced by instrumentation, and not
about the logs produced by the application.

```

> Logging level is controlled by an ENV variable `OTEL_LOG_LEVEL`, two most common values are:
> - `INFO` - the default value
> - `VERBOSE` - highest value likely to be needed
> 

Reviewer: ```suggestion
Logging level is controlled by the `OTEL_LOG_LEVEL` environment variable. The two most common values are:
- `INFO`: the default value
- `VERBOSE`: highest value likely to be needed

```

For all possible log levels see
[this source file](https://github.com/open-telemetry/opentelemetry-js-api/blob/main/src/diag/types.ts).

> There is no default output for logs. Even if you set `OTEL_LOG_LEVEL=VERBOSE`, there won't be anything in the console.
> You need to set an output first, for example to `stdout`, by adding `DiagConsoleLogger`:

Reviewer: ```suggestion
There is no default output for logs. Even if you set `OTEL_LOG_LEVEL=VERBOSE`, nothing is output to the console. You need to set an output first, for example to `stdout`, by adding `DiagConsoleLogger`:
```

# Migrate from the SignalFx Tracing Library for NodeJS

The Splunk Distribution of OpenTelemetry for NodeJS replaces the SignalFx Tracing
Library for NodeJS. If you’re using the SignalFx Tracing Library, migrate to
the Splunk Distribution of OpenTelemetry for NodeJS to use OpenTelemetry
instrumentation to send traces to Splunk APM. The Splunk Distribution of
OpenTelemetry for NodeJS uses OpenTelemetry to instrument applications, which is
an open-source API to gather telemetry data, and has a smaller footprint.

Because the SignalFx Tracing Library for NodeJS uses OpenTracing and the Splunk Distribution
of OpenTelemetry for NodeJS uses OpenTelemetry, the semantic

> conventions for span tag names change when you migrate. For more information,

Reviewer 2: .. and the span names as well! There are different guidelines for that in OTel.

see [Migrate from OpenTracing to OpenTelemetry](https://docs.signalfx.com/en/latest/apm/apm-getting-started/apm-opentelemetry-collector.html#apm-opentelemetry-migration).

## Getting help

If you experience any issues following the guide below, or something is unclear, or missing, please don't hesitate to
open an issue in Github. Any and all ideas for improvements are also welcome.

<a name="known-limitations"></a>
## Known limitations as compared to SignalFx Tracing Library

- Lowest supported version of NodeJS is `v8.5`, [see more information here](https://github.com/open-telemetry/opentelemetry-js#node-support)
- No auto-instrumentation for:
  - `AdonisJS`
  - `amqp10`

>   - `mongodb-core` ([because it's deprecated](https://github.com/mongodb-js/mongodb-core))

Reviewer 2: I would stress that MongoDB driver auto instrumentation is still supported. They just reogranized the client lib.

  - `sails`
- Limited instrumentation for:
  - `nest` - only manual insturmentation helpers, provided by community
- other notes on instrumentation:

>   - `express` instrumentation requires `http`/`https` instrumentation

Reviewer 2: Same applies for koa and hapi for now.

Reviewer 2: Oh... and it's duplicated few lines below.

  - `bluebird` - context propagation only (via `async_hooks`)

>   - `q` - context propagation only (via `async_hooks`)

Reviewer 2: Dunno if I'd call "context prop only" a limitation for a Promise lib. I don't personally think anyone expects to instrument all the async resources in their application.

Author: Yeah, it was already brought up a couple times, I'll reword it.

Author: Yeah, it was already brought up a couple times, I'll reword it.

  - `when` - context propagation only (via `async_hooks`)
  - `socket.io` - provided by community (<https://github.com/aspecto-io/opentelemetry-ext-js/tree/master/packages/instrumentation-socket.io>)

## Changes to defaults

> - Default flush interval, which defines how frequently captured telemetry data is sent to the backend, is now 30s instead of 2s

Reviewer 2: I'd add a link to docs to change that or state "to change that back to 2s, set ENVVAR=2000" or something.

EDIT

OK, got to the section at the bottom and discovered that it's not configurable out of the box. Maybe make that configurable in the future? I'd make an issue about that and ignore my comment above.

Author: I added #198, I honestly don't know how soon we'll get to it.

## Requirements

> This Splunk Distribution of OpenTelemetry requires Node.js 8.5 or later.

Reviewer 2: That's also stated under limitations. I'd leave one or the other.

Author: I'll simplify this.

If you're still using an earlier version of Node.js, continue using the SignalFx Tracing Library for Node.js.

Current effective required Node.js version is: ![node-current](https://img.shields.io/node/v/@splunk/otel?style=flat-square)

## Migration steps

### Instrumented libraries

With the exception of [explicitly listed limitations](#known-limitations) we aim to support all libraries supported by
signalfx-nodejs-tracing. To find an equivalent auto-instrumentation open <https://opentelemetry.io/registry/> and for
each instrumentation
[from `signalfx-nodejs-tracing`'s README](https://github.com/signalfx/signalfx-nodejs-tracing/#requirements-and-supported-software)
search by the name of the library in the registry.

For example, if you'd like to migrate instrumentation for `mysql`, go to
[https://opentelemetry.io/registry/?s=**mysql**&component=&language=**js**#](https://opentelemetry.io/registry/?s=mysql&component=&language=js#).

Once you have identified an instrumentation package, install it using `npm` or `yarn`.

- If the package is [on this list](./README.md#default-instrumentation-packages-), it

>   will be enabled automatically.

Reviewer 2: but not installed automatically, right? I'd stress that somehow or make that more clear unless you think I'm the only one that had to stop here and think.

| SIGNALFX_ENDPOINT_URL              | n/a                                  | OTLP is not implemented yet |
| SIGNALFX_RECORDED_VALUE_MAX_LENGTH | SPLUNK_MAX_ATTR_LENGTH               |       |
| SIGNALFX_TRACING_DEBUG             | no direct equivalent                 | see [instrumentation logs](#instrumentation-logs) |
| SIGNALFX_SPAN_TAGS                 | OTEL_RESOURCE_ATTRIBUTES             | format needs to be changed to `key1=val1,key2=val2` |
| SIGNALFX_LOGS_INJECTION            | SPLUNK_LOGS_INJECTION                | |
| SIGNALFX_LOGS_INJECTION_TAGS       | OTEL_RESOURCE_ATTRIBUTES             | there's no direct equivalent, but values specified in `OTEL_RESOURCE_ATTRIBUTES` will also be used for logs injection |
| SIGNALFX_ENABLED_PLUGINS           | n/a                                  | see [the README section about instrumentations](./README.md#custom-instrumentation-packages) |
| SIGNALFX_SERVER_TIMING_CONTEXT     | SPLUNK_TRACE_RESPONSE_HEADER_ENABLED | |
| SIGNALFX_TRACING_ENABLED           | OTEL_TRACE_ENABLED                   | |

### Programmatic configuration

Update these programmatic configuration options:

> | Old property             | New property            | Notes |

Reviewer 2: Instead of "old" and "new" I'd use the names "OpenTracing" and "OpenTelemetry" or `signalfx-tracing` and `@splunk/otel`.

| `accessToken`            | `accessToken`           |       |
| `enabled`                | -                       | no equivalent, but Environment Variable can be used |
| `debug`                  | -                       | no direct equivalent, see [instrumentation logs](#instrumentation-logs) |
| `tags`                   | `tracerConfig.resource` | |
| `logInjection`           | `logInjectionEnabled`   | |
| `logInjectionTags`       | -                       | no direct equivalent, but `tracerConfig.resource` can be used |
| `flushInterval`          | -                       | no direct equivalent, contact us if you had customized this value |
| `plugins`                | -                       | see [the README section about instrumentations](./README.md#custom-instrumentation-packages) |
| `recordedValueMaxLength` | `maxAttrLength`         | |
| `enableServerTiming`     | `serverTimingEnabled`   | |

### Instrumentation entry point

```javascript
const tracer = require('signalfx-tracing').init({
  // your options here
})
```

becomes

```javascript
const { startTracing } = require('@splunk/otel');

startTracing({
  // your new options here
});
```

> and requires installing `@splunk/otel` first, using either NPM or Yarn. Same as in `signalfx-nodejs-tracing`, this code

Reviewer 2: npm is always written lowercase.

is to be run before your `import` or `require` statements.

Alternatively, you can append the flag `-r @splunk/otel/instrument` instead when launching `node` (it runs
`startTracing` under the hood). In that case you cannot use programmatic configuration and must rely on environment
variables only.

### Instrumentation logs

There isn't a one-to-one mapping for `SIGNALFX_TRACING_DEBUG`. The closest equivalent is `OTEL_LOG_LEVEL`, however the logged

> information might be different.

Reviewer 2: **will** be different.
